File: fp-java-lib/src/main/java/io/funfun/redbook/Stream.java
package io.funfun.redbook;

import java.util.NoSuchElementException;
import java.util.function.Predicate;
import java.util.function.Supplier;

public abstract class Stream<T> {

    private final static Stream NONE = new None();

    abstract T head();
    abstract Stream<T> tail();
    abstract boolean isEmpty();

    public static <T> Supplier<Stream<T>> none() {
        return (() -> NONE);
    }

    public static <T> Stream<T> seq(Supplier<T> head, Supplier<Stream<T>> tail) {
        return new Seq<>(head, tail);
    }

    public static <T> Stream<T> stream(T... args) {
        Stream<T> result = NONE;
        for (int i = args.length - 1; i >= 0; i--) {
            final Stream<T> tail = result;
            final T head = args[i];
            result = seq(() -> head, () -> tail);
        }
        return result;
    }

    public static <T> Stream<T> stream(Supplier<T>... args) {
        Stream<T> result = NONE;
        for (int i = args.length - 1; i >= 0; i--) {
            final Stream<T> tail = result;
            result = seq(args[i], () -> tail);
        }
        return result;
    }

    public <T> Stream<T> filter(Predicate<T> predicate) {
        return this.filter(NONE, predicate);
    }

    protected  <T> Stream<T> filter(Stream<T> acc, Predicate<T> predicate) {
        if (this instanceof None) {
            return acc.reverse();
        } else {
            if (predicate.test(((Seq<T>)this).head())) {
                return this.tail().filter(new Seq(() -> ((Seq<T>)this).head(), () -> acc), predicate);
            } else {
                return this.tail().filter(acc, predicate);
            }
        }
    }

    public <T> Stream<T> reverse() {
        return reverse(NONE);
    }

    protected <T> Stream<T> reverse(Stream<T> acc) {
        if (this instanceof None)
            return acc;
        else {
            return this.tail().reverse(new Seq(() -> ((Seq<T>)this).head(), () -> acc));
        }
    }

    public static class Seq<T> extends Stream<T> {

        private final Supplier<T> head;
        private final Supplier<Stream<T>> tail;
        private T _head;
        private Stream<T> _tail;

        public Seq(Supplier<T> head, Supplier<Stream<T>> tail) {
            this.head = head;
            this.tail = tail;
        }

        @Override
        public T head() {
            if (_head == null) {
                _head = head.get();
            }
            return _head;
        }

        @Override
        public Stream<T> tail() {
            if (_tail == null) {
                _tail = tail.get();
            }
            return _tail;
        }

        @Override
        boolean isEmpty() {
            return false;
        }

        @Override
        public String toString() {
            return "Seq(\"" + head() + "\", " + tail() + ")";
        }
    }

    public static class None extends Stream<Object> {

        @Override
        public Object head() {
            throw new NoSuchElementException();
        }

        @Override
        public Stream<Object> tail() {
            throw new NoSuchElementException();
        }

        @Override
        boolean isEmpty() {
            return true;
        }

        @Override
        public String toString() {
            return "None";
        }
    }
}

File: fp-kotlin-lib/src/main/kotlin/io/funfun/redbook/Stream.kt
package io.funfun.redbook

sealed class Stream<out T> {

    companion object {
        operator fun <T>invoke(vararg values: () -> T): Stream<T> {
            var results: Stream<T> = None
            for (value in values.reversed()) {
                val newTail = results
                results = Seq( value, { newTail })
            }
            return results
        }
    }
}

data class Seq<T>(private val _head: () -> T, val _tail: () -> Stream<T>): Stream<T>() {
    var head: T? = null
        private set
        get() = field?: _head.invoke().also {
            field = it
        }
    var tail: Stream<T>? = null
        private set
        get() = field?: (_tail.invoke()).also {
            field = it
        }
}
object None: Stream<Nothing>()

fun <T> Stream<T>.head(): T = when {
    this is Seq -> head ?: Unit as T
    else -> Unit as T
}

fun <T> Stream<T>.tail(): Stream<T> = when {
    this is Seq -> tail ?: None
    else -> None
}

fun <T> Stream<T>.filter(acc: Stream<T> = None, predicate: (T) -> Boolean): Stream<T> = when(this) {
    is None -> acc
    is Seq -> if (predicate(head())) {
        head().let {
            tail.filter(Seq({ it }, { acc }), predicate)
        }
    } else {
        tail.filter(acc, predicate)
    }
}

tailrec fun <T> Stream<T>.reverse(acc: () -> Stream<T> = { None }): Stream<T> = when(this) {
    is None -> acc()
    is Seq -> tail().reverse { Seq({ head() }, acc)}
}

// 성능상 이게 좋음, 직관적인 면에서는 foldRight가 좋음
tailrec fun <T, R> Stream<T>.map(acc: (() ->Stream<R>) = { None }, transformer: (T) -> R): Stream<R> = when(this) {
    is None -> acc().reverse()
    is Seq -> tail.map({ Seq({transformer(head())}, acc) }, transformer)
}

tailrec fun <T, R> Stream<T>.foldLeft(z: R, op: (R, T) -> R): R = when(this) {
    is None -> z
    is Seq -> tail().foldLeft(op(z, head()), op)
}

fun <T, R> Stream<T>.foldRight(z: R, op: (T, R) -> R): R = when(this) {
    is None -> z
    is Seq -> op(head(), tail().foldRight(z, op))
}